sparsecoding/transforms/whiten.py

```diff
@@ -1,136 +1,114 @@
 import torch
+from typing import Dict
 
 
-class Whitener(object):
-    """Performs data whitening (and un-whitening).
+def compute_whitening_stats(X: torch.Tensor):
 
-    On initialization, stores the mean and covariance of the input data.
-    These statistics are then used for future `whiten()` or `unwhiten()` calls.
+    """
+    Given a tensor of data, compute statistics for whitening transform.
 
     Parameters
     ----------
-    data : Tensor, shape [N, D]
-        Data to be whitened,
-        where:
-            N is the number of data points,
-            D is the dimension of the data points.
+    X: Input data of size [N, D]
 
-    Attributes
+    Returns
     ----------
-    mean : Tensor, shape [D]
-        Mean of the input data for the Whitener.
-    covariance : Tensor, shape [D, D]
-        Co-variance of the input data for the Whitener.
-    eigenvalues : Tensor, shape [D]
-        Eigenvalues of `self.covariance`.
-    eigenvectors : Tensor, shape [D, D]
-        Eigenvectors of `self.covariance`.
-    epsilon : float
-        Prevents division by zero.
+    Dictionary containing whitening statistics (eigenvalues, eigenvectors, mean)
+    """
+
+    mean = torch.mean(X, dim=0)
+    X_centered = X - mean
+    Sigma = torch.cov(X_centered.T)
+
+    eigenvalues, eigenvectors = torch.linalg.eigh(Sigma)
+
+    # Since eigh returns values in ascending order, reverse them to get descending order
+    eigenvalues = torch.flip(eigenvalues, dims=[0])
+    eigenvectors = torch.flip(eigenvectors, dims=[1])
+
+    return {
+        'mean': mean,
+        'eigenvalues': eigenvalues,
+        'eigenvectors': eigenvectors,
+        'covariance': Sigma
+    }
+
+
+def whiten(X: torch.Tensor,
+           algorithm: str = 'zca',
+           stats: Dict = None,
+           n_components=None,
+           epsilon: float = 0.
+           ) -> torch.Tensor:
+
+    """
+    Apply whitening transform to data using pre-computed statistics.
+
+    Parameters
+    ----------
+    X: Input data of shape [N, D] where N are unique data elements of dimensionality D
+    algorithm: Whitening transform we want to apply, one of ['zca', 'pca', or 'cholesky']
+    stats: Dict containing precomputed whitening statistics (mean, eigenvectors, eigenvalues)
+    n_components: Number of principal components to keep. If None, keep all components.
+                  If int, keep that many components. If float between 0 and 1,
+                  keep components that explain that fraction of variance.
+    epsilon: Optional small constant to prevent division by zero
+
+    Returns
+    ----------
+    Whitened data of shape [N, D]
+
+    Notes
```

Comment:
I think it would be helpful for documentation to
- add a few words (super short) for why someone would want to choose PCA or Cholesky over the the default ZCA, and then direct them to the paper or stackexchange for more details.
- add short example usage in a code block that calls `compute_whitening_stats` and then `whiten` (not sure how to format this for Sphinx @belsten). Maybe this is excessive, but I think we should be clear because these are separate functions. Or could mention first calling `compute_whitening_stats` in the `stats` Parameter field.

What do you think?

Reply:
I cover all of that in the whitening notebook that'll be added to the documentation. What about adding that notebook to the references to encourage people to check it out?

Reply:
A pointer to the notebook would work instead IMO! And it avoids bloat in the docstring.

Reply:
Done!